internal/pkg/agent/application/upgrade/upgrade.go

```diff
@@ -170,49 +194,126 @@ func (u *Upgrader) Upgrade(ctx context.Context, version string, sourceURI string
 
 	det.SetState(details.StateExtracting)
 
-	newHash, err := u.unpack(version, archivePath, paths.Data())
+	metadata, err := u.getPackageMetadata(archivePath)
+	if err != nil {
+		return nil, fmt.Errorf("reading metadata for elastic agent version %s package %q: %w", version, archivePath, err)
+	}
+
+	currentVersion := agentVersion{
+		version:  release.Version(),
+		snapshot: release.Snapshot(),
+		hash:     release.Commit(),
+	}
+
+	newVersion := agentVersion{}
+	if metadata.manifest != nil {
+		packageDesc := metadata.manifest.Package
+		newVersion.version = packageDesc.Version
+		newVersion.snapshot = packageDesc.Snapshot
+	} else {
+		// extract version info from the version string (we can ignore parsing errors as it would have never passed the download step)
+		parsedVersion, _ := agtversion.ParseVersion(version)
+		newVersion.version = strings.TrimSuffix(parsedVersion.VersionWithPrerelease(), snapshotSuffix)
+		newVersion.snapshot = parsedVersion.IsSnapshot()
+	}
+	newVersion.hash = metadata.hash
+
+	u.log.Debugw("Comparing current and new agent version", "current_version", currentVersion, "new_version", newVersion)
+
+	if currentVersion == newVersion {
+		return nil, fmt.Errorf("agent version is already %s", currentVersion)
+	}
+
+	u.log.Infow("Unpacking agent package", "version", newVersion)
+
+	// Nice to have: add check that no archive files end up in the current versioned home
+	unpackRes, err := u.unpack(version, archivePath, paths.Data())
 	if err != nil {
 		return nil, err
 	}
 
+	newHash := unpackRes.Hash
 	if newHash == "" {
 		return nil, errors.New("unknown hash")
 	}
 
-	if strings.HasPrefix(release.Commit(), newHash) {
-		u.log.Warn("Upgrade action skipped: upgrade did not occur because its the same version")
-		return nil, nil
+	if unpackRes.VersionedHome == "" {
+		// FIXME this should be treated as an error
```

Comment:
It looks like it is an error already?

Reply:
Yeah, decided for error and then forgot to remove the fixme, will remove that